### DISCUSSION TITLE
'ComboBoxState<T>' is assignable to the constraint of type 'T', but 'T' could be instantiated with a different subtype of constraint 'object'

### DISCUSSION THREAD

**Discussion by mgo-af** [adobe/react-spectrum#4456] [Q&A]
Hello!

Just running through the `react-aria-components` examples, I run into this TS error when copy-pasting the example into a sandbox.

```
Type 'ReactNode | ((item: T) => ReactNode)' is not assignable to type 'ReactNode | ((item: T) => ReactElement<any, string | JSXElementConstructor<any>>)'.
  Type '(item: T) => ReactNode' is not assignable to type 'ReactNode | ((item: T) => ReactElement<any, string | JSXElementConstructor<any>>)'.
    Type '(item: T) => ReactNode' is not assignable to type '(item: T) => ReactElement<any, string | JSXElementConstructor<any>>'.
      Type 'ReactNode' is not assignable to type 'ReactElement<any, string | JSXElementConstructor<any>>'.
        Type 'string' is not assignable to type 'ReactElement<any, string | JSXElementConstructor<any>>'.ts(2322)
types.d.ts(50, 5): The expected type comes from property 'children' which is declared here on type 'IntrinsicAttributes & ListBoxProps<T> & RefAttributes<HTMLDivElement>'
```

[Here is a sandbox](https://codesandbox.io/s/purple-tdd-ccfpp4?file=/src/MyCombobox.tsx)

**Comment by LFDanLu** [comment]
Good catch, I believe the docs are wrong here and that the `MyComboBoxProps` type should define children as `children: React.ReactNode | ((item: T) => React.ReactElement);` like [so](https://codesandbox.io/s/kind-sammet-vl3s7e?file=/src/MyCombobox.tsx:388-452). This is because the  modified `MyComboBox` wants to provide children directly to the ListBox which has a definition of `children?: ReactNode | ((item: T) => ReactElement)` (aka the common children definition for a Collection element). The original ComboBoxProps doesn't follow this collection children definition, instead accepting RenderProps instead.

**Reply by LFDanLu** [reply]
Opened https://github.com/adobe/react-spectrum/issues/4470 to track this
